apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl

```diff
@@ -570,17 +904,67 @@ session_iterator_next(Cursor0, N) ->
     ],
     {NextVals, Cursor} = session_iterator_next(mnesia:dirty_next(?session_tab, Cursor0), N - 1),
     {ThisVal ++ NextVals, Cursor}.
+%% END ifdef(STORE_STATE_IN_DS).
+-endif.
 
 %%================================================================================
 %% Internal functions
 %%================================================================================
 
 %% All mnesia reads and writes are passed through this function.
 %% Backward compatiblity issues can be handled here.
+-ifdef(STORE_STATE_IN_DS).
+val_encode(_Domain, Term) ->
+    term_to_binary(Term).
+
+val_decode(_Domain, Bin) ->
+    binary_to_term(Bin).
+
+-spec key_encode(domain(), term()) -> binary().
+key_encode(?metadata_domain, _Key) ->
+    ?metadata_domain_bin;
+key_encode(?subscription_domain, TopicFilterAndSubId) ->
+    term_to_topic_level(TopicFilterAndSubId);
+key_encode(?subscription_state_domain, SubStateId) ->
+    integer_to_binary(SubStateId);
+key_encode(?stream_domain, StreamKey) ->
+    term_to_topic_level(StreamKey);
+key_encode(?rank_domain, RankKey) ->
+    term_to_topic_level(RankKey);
+key_encode(?seqno_domain, SeqnoType) ->
+    integer_to_binary(SeqnoType);
+key_encode(?awaiting_rel_domain, PacketId) ->
+    integer_to_binary(PacketId).
+
+-spec key_decode(domain(), binary()) -> term().
+key_decode(?metadata_domain, Bin) ->
+    Bin;
+key_decode(?subscription_domain, Bin) ->
+    topic_level_to_term(Bin);
+key_decode(?subscription_state_domain, Bin) ->
+    binary_to_integer(Bin);
+key_decode(?stream_domain, Bin) ->
+    topic_level_to_term(Bin);
+key_decode(?rank_domain, Bin) ->
+    topic_level_to_term(Bin);
+key_decode(?seqno_domain, Bin) ->
+    binary_to_integer(Bin);
+key_decode(?awaiting_rel_domain, Bin) ->
+    binary_to_integer(Bin).
+
+term_to_topic_level(Term) ->
+    base64:encode(term_to_binary(Term), #{mode => urlsafe, padding => false}).
```

Comment:
Do we need base64? Since these "topics" probably won't be exposed via MQTT protocol, we don't care if topic level is valid UTF8.

Reply:
It was a way to ensure that the topic level doesn't contain invalid characters like `/`, `+`, `#`, whatever the input is.  We can use another encoding that satisfies that, sure.

Reply:
Fair enough.

Reply:
On a second thought, DS API treats topics as lists of binaries, and wildcards in the topic filters as atoms `+` and `#`. There should not be any `binary()` -> `[word()]` parsing.
The only exception is `skipstream_lts` layout's topic compaction... It does not have escaping for `/`'s in the topic levels currently. But that can be fixed.

Reply:
> On a second thought, DS API treats topics as lists of binaries, and wildcards in the topic filters as atoms + and #. There should not be any binary() -> [word()] parsing.

I think I'm a bit lost, sorry.  Where would that binary() -> [word()] parsing be involved in?

The highlighted `term_to_topic_level` function was used for encoding arbitrary terms (such as `{[word()], nil()}`, for example) to a single topic level.  We need to encode to binary and then escape bad characters (like `#`) because we need to project terms to a binary topic level.